### DISCUSSION TITLE
Feature Request: SSH agent forward key settings

### DISCUSSION THREAD

**Discussion by mangas** [blinksh/blink#1508] [Ideas]
I would like to request an option on the host to enable Agent forwarding by default with the key already added to it. 

Currently it is necessary to call ssh-add KEY_NAME and then ssh -a HOST. 

Would be great to just either have the relevant ssh key added to the agent by default or have an option per host to select keys to forward. (forwarding all is also ok, prolly worth having a setting to disable it).

**Comment by carloscabanero** [comment]
Last time I took a look at this I blocked it because we wanted to do it following what the standard .ssh/config file format would dictate. There is the AddKeysToAgent flag, and that would add the IdentityFile keys (https://man7.org/linux/man-pages/man5/ssh_config.5.html). I’m not completely sold on that, as I usually separate my keys and do not want a “login” key to then make it to the other side in a “generic” way.

We could provide our own interface and corresponding config file to even have different “agents” for different sets of hosts, etc…

At the same time, there are changes coming up to the OpenSSH Agent itself (ie https://www.openssh.com/agent-restrict.html), so for now we thought it may make sense to wait and see what they come up with as well.

Please leave us your thoughts!

**Comment by mangas** [comment]
@carloscabanero I've tried to set AddKeyToAgent on the host config and it didn't work, I see the key in `ssh-add -l` but the agent doesn't sign the requests. 

On the point of having finer control over the keys, I think the app should not impose the use of a "login" key and make mandatory to have it available on the host. I think the main goal here is to provide a more efficient flow for a specific use case. You'd still be able to use your flow while supporting a simpler flow for whoever wants to buy into it. 

In any case I'm happy to use a configuration like `AddKeysToAgent` and I agree sticking to the standards would be ideal although so far I've not been able to configure the keys for automatic import, always need to add the keys manually before ssh.

**Reply by carloscabanero** [reply]
We have improved this now on 17.3.0, the default agent now has a special configuration. Please check out https://docs.blink.sh/advanced/advanced-ssh#ssh-agent-and-agent-forwarding For more information!

**Comment by mhoran** [comment]
When the default agent is configured to confirm signature requests, this also applies to connections initiated by Blink. It seems that setting up a default agent effectively makes the agent used by Blink itself. That is how SSH behaves on my computer as well; but I did not expect to have to confirm signature requests made by Blink.

I think if `IdentitiesOnly` were available as a configuration option, then this issue could be mitigated. However, perhaps Blink should not require confirmation to use the agent itself? Or, the configuration and alert dialog should be clarified, as this really confused me. I thought I was being asked for confirmation to enable forwarding (which is what the message says), but if I click "Do not forward", then I get `Error connecting to host. authError(msg: "ssh_userauth_agent_publickey: Out of memory")`.

If I set a per-connection agent, then it behaves as I expect. The agent is *not* used for the connection itself, but is made available to the other side of the connection, and confirmation is requested when something on the other side tries to use the agent. However, I could not find a way to set up a per-connection agent via `~/.ssh/config`, only via the UI.

I don't generally want to run an agent or forward, but sometimes it is necessary -- hence I like to lock it down as much as possible, as per the Blink documentation.
